### ISSUE TITLE
Docker Workflow Keeps Failing on PRs

### ISSUE THREAD

**Issue by ThunderSon** [tcort/markdown-link-check#136]
After sending my PR, I reviewed that the failure of the Docker publishing happened on the PRs.
It's because the token isn't being shared with the destination repository.

I'd advise to remove the following from the workflow:
```
  pull_request:
    paths-ignore:
      - '.**'
```
Or otherwise, have a workflow tweak that checks if the container can be deployed locally on PRs, and then you publish on merge.

**Comment by NicolasMassart** [comment]
@cadeef What would be the downside of not publishing the Docker image on PRs?
The token is only for pushing to the GH registry, isn't it?
Only building the image for PR would be enough to make sure it works.
Otherwise it requires to use [pull_request_target](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target) for what I understood and honestly I'm not GH Action expert enough to understand how... But I would love to learn 👍

**Comment by cadeef** [comment]
Hmm... I'm not sure I understand the context, but here's what I think:

I assume we're talking about workflows on forked repos? GH Actions don't run by default on forks, has the function changed? On mobile, but can check later if that's the case. That's the only situation I can think of where lack of personal access token from the upstream repo would make a difference.

I agree that tweaking the run conditions of CI actions would be a good idea. I'm not sure about this particular suggestion though... what does it accomplish? Admittedly, when I added the filter initially it was out of convenience and intended to stop image rebuilds on non-core (non-software) changes.

It would probably be beneficial to "whiteboard" out the flow for current/future CI to get the waterfall correct and minimize unneeded work.

**Comment by ThunderSon** [comment]
So currently the PR is not activating the workflow successfully because of the token.
To summarize the difference between PR and PRT:
- `pull_request` -> grab the token from the source repository (tcort in this case)
- `pull_request_target` -> grab the token from the destination repository (thunderson in this case)
  - This should be used for security reasons (secrets leak)

To access secrets, you need a certain scope to do that. More on that in [here](https://developer.github.com/v3/actions/secrets/#list-repository-secrets).

Whenever a PR is merged, the publishing works on the "merge commit".
What you could benefit from if a PR is tested is the container build success (I was able to build the container, and have my tests valid, ok good this PR won't break my container). That's the risk you're running now, that a merge commit fails and doesn't deploy the latest version.

**Comment by cadeef** [comment]
Ok, I follow now, thanks for the write up @ThunderSon.

https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/ might be helpful for others too.

I think you're mostly right, and since the PR triggered images are mostly a convenience feedback item it is worth considering their impact. I'll let it percolate for a bit (it's obviously not mine to decide either, but I've done a bit of the work).

**Comment by NicolasMassart** [comment]
I agree with both of you on this one: making sure image is ok in as soon as possible is the way to go.
But only the build is necessary in the PR IMO. So do we need the toke if we only enable the build and not the push to a registry? Push to a registry is not very useful in a PR except if we want to test the image build in the PR... but that's probably once in a while and not the golden path for the moment.

**Comment by cadeef** [comment]
I know how we can make it work. I'll write it up, but don't have the bandwidth for it today.

**Comment by ThunderSon** [comment]
I don't mind reviewing if you're up for the task. Otherwise, let me know if I can help in any other way.
